web/src/recent_view_data.ts

```diff
@@ -21,14 +46,40 @@ export function process_message(msg: Message): boolean {
     let conversation_data_updated = false;
 
     // Initialize conversation data
-    const key = get_key_from_message(msg);
+    const key = get_key_from_conversation_data(msg);
     let conversation_data = conversations.get(key);
     if (conversation_data === undefined) {
-        conversation_data = {
-            last_msg_id: -1,
-            participated: false,
-            type: msg.type,
-        };
+        const participated = false;
+        const last_msg_id = -1;
+        const latest_mesage_timestamp = msg.timestamp;
+        const unread_count = message_to_conversation_unread_count(msg);
+
+        if (msg.type === "private") {
+            // Display recipient contains user information for DMs.
+            assert(typeof msg.display_recipient !== "string");
```

Comment:
This is just preserving old logic, but ideally we'd refactor the message type so that we don't have to make this assertion.

Reply:
So, we need to also think about how we are going to update the properties here which can change.

Like we used to update `unread_count` by simply calling `inplace_rerender` since that would get the latest unread count from `last_msg_id` and update it, but that won't work anymore since we using a cached value.

We need to hook server_events for this. We could use the `process_message` as an opportunity to update them but that won't be reliable to probably a good idea to skip updating them here.

Reply:
Agreed with @amanagr that we're going to need to be very careful to make sure things that could change any of the data that you're caching here, which is quite a lot (name changes go into `display_recipient`, etc.).

Realistically that means this is the wrong design; we should be only caching the metadata that we think is stable, not computed properties like `display_recipient`, which can just as well be processed later.

Reply:
Just to lay out a bit more how I'm thinking this should work: A server API endpoint that sent data on what recent conversations exist would not include the following:
- unread data -- clients should combine its unread data set with the recent view data to display a complete view.
- Metadata on users, like names/etc. that go into display recipients -- I'd expect it to just have stream/user IDs.

Instead, I'd expect it to just have (1) list of participated user IDs, (2) `last_message_id`, (3) `last_message_timestamp`, (4) type and location of the message.

So I think it's correct for the data structure we're defining here to have those properties as well, and rely on combining that reasonably normalized data set with the `unread` data set and `people/stream_data` data sets, just like it will with which topics are followed in order to display that detail.

Reply:
I just pushed to this branch with an extra commit to try addressing this -- let me know what you think! I wasn't totally sure which data we can rely on to be stable.